clippy_lints/src/static_mut.rs

```diff
@@ -0,0 +1,60 @@
+use clippy_utils::diagnostics::span_lint_and_help;
+use rustc_ast::ast::{Item, ItemKind, Mutability, StaticItem};
+use rustc_lint::{EarlyContext, EarlyLintPass, LintContext};
+use rustc_middle::lint::in_external_macro;
+use rustc_session::declare_lint_pass;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Produces warnings when a `static mut` is declared.
+    ///
+    /// ### Why is this bad?
+    /// `static mut` can [easily produce undefined behavior][1] and
+    /// [may be removed in the future][2].
+    ///
+    /// ### Example
+    /// ```no_run
+    /// static mut GLOBAL_INT: u8 = 0;
+    /// ```
+    /// Use instead:
+    /// ```no_run
+    /// use std::sync::RwLock;
+    ///
+    /// static GLOBAL_INT: RwLock<u8> = RwLock::new(0);
+    /// ```
+    ///
+    /// [1]: https://doc.rust-lang.org/nightly/edition-guide/rust-2024/static-mut-reference.html
+    /// [2]: https://github.com/rust-lang/rfcs/pull/3560
+    #[clippy::version = "1.80.0"]
+    pub STATIC_MUT,
+    nursery,
+    "detect mutable static definitions"
+}
+
+declare_lint_pass!(StaticMut => [STATIC_MUT]);
+
+impl EarlyLintPass for StaticMut {
+    fn check_item(&mut self, cx: &EarlyContext<'_>, item: &Item) {
```

Comment:
The issue mentioned some suggestions to replace `static mut` including using `Mutex` or `Atomic`s, but the problem is non of those are available in `no_std` context, so it might be a good idea to check that with [`is_no_std_crate`](https://doc.rust-lang.org/nightly/nightly-rustc/clippy_utils/fn.is_no_std_crate.html). But you might need to change the pass to `Late`

Reply:
The issue does mention `SyncUnsafeCell` (or the stable alternative: a newtype around `UnsafeCell` with an `unsafe impl Sync`). I think that should work fine even in `no_std` since they're defined in `core`. The description should probably mention this though.